libseccomp-sys/build.rs

```diff
@@ -7,24 +7,203 @@ use std::env;
+        // Technically, we don't have to do this and can extract it from `libseccomp/src/.libs/` after
+        // the build is done but it's cleaner this way.
+        //
+        // If the directory already exists, we delete it and start over. This should only
+        // happen if the build is being rerun because the environment variables changed, or the
+        // build itself failed.
+        let mut build_output_path = out_dir;
+        build_output_path.push("libseccomp-build-out");
+
+        if build_output_path.exists() {
+            std::fs::remove_dir_all(&build_output_path)?;
+        }
+        std::fs::create_dir(&build_output_path)?;
+
+        // Build source with configure, make, make install, setting the output
+        // directory to a directory we'll make in our package's out_dir.
+
+        // Get compiler flags info.
+        let cc_info = cc::Build::new().get_compiler();
+        let cc_env = cc_info.cc_env();
+        let cflags_env = cc_info.cflags_env();
+
+        // TODO: maybe could replace these commands with just using cc if the makefile isn't doing
+        // anything significant?
+        let configure_result = std::process::Command::new("./configure")
+            .current_dir(&src_path)
+            .args(["--prefix", build_output_path.to_str().unwrap()])
+            .arg(format!("CFLAGS={}", cflags_env.to_str().unwrap()))
+            .arg(format!("CC={}", cc_env.to_str().unwrap()))
+            .spawn()?
+            .wait()?;
+        assert!(
+            configure_result.success(),
+            "Running `configure` failed while building libseccomp."
+        );
+
+        let make_result = std::process::Command::new("make")
+            .current_dir(&src_path)
+            .spawn()?
+            .wait()?;
+        assert!(
+            make_result.success(),
+            "Running `make` failed while building libseccomp."
+        );
+
+        let make_install_result = std::process::Command::new("make")
+            .current_dir(&src_path)
+            .arg("install")
+            .spawn()?
+            .wait()?;
+        assert!(
+            make_install_result.success(),
+            "Running `make install` failed while building libseccomp."
+        );
+
+        // Return the lib path.
+        build_output_path.push("lib");
+
+        Ok(build_output_path)
+    }
 
-    Ok(())
+    // NOTE: We could replace this with reqwest, flate2, and sha2 but it would increase the
+    // build-time compile time (on the first run), in exchange for not having to have those
+    // commands installed on the machine building the crate.
+    /// Uses curl and tar and sha256sum to download, extract, and verify the libseccomp source.
+    fn download_libseccomp(out_dir: &Path) -> Result<PathBuf, Box<dyn std::error::Error>> {
+        let url = LIBSECCOMP_RELEASE_URL.replace("VERSION", LIBSECCOMP_VERSION);
+        let tarball_name = "libseccomp-VERSION.tar.gz".replace("VERSION", LIBSECCOMP_VERSION);
```

Comment:
I'm okay to do this, but it seems also possible to do this by putting the values into a cargo config at `libseccomp-sys/.cargo/config.toml` and then reading the values from environment variables.

The benefit is we don't have to include extra dependencies to read from configuration files, but the main drawback is that the file is a bit hidden and easy to forget. This could be mitigated by including the details of upgrading the bundled libseccomp version in a "Maintenance" section in the readme, and making sure to leave a comment at the point of use about where the environment variable is set.

 (see [the cargo config documentation](https://doc.rust-lang.org/nightly/cargo/reference/config.html#env)] for more information)

Reply:
TBH a ini-like parser is written in a few lines like (not finish):

```rust
fn from_str(s: &str) -> HashMap<String, HashMap<String, String>> {
    let mut versions = HashMap::new();
    let mut cur_version;
    for line in s.lines() {
        let line = line.trim();
        if line.is_empty() || line.starts_with("#") {
            continue;
        }
        
        if line.starts_with("[") {
            cur_version = versions
                .entry(line.trim_matches(&['[', ']'][..]).to_string())
                .or_default();
        } else {
            if let Some((k, v)) = line {
                cur_version.insert(k, v)
            }
        }
    }
    versions
}
```

IMHO we don't need more than a simple key=value format.

Reply:
BTW we can also manage this in a 

versions.rs

```rust
static LIBSECCOMP: SourceInfo = SourceInfo {
    description: "High level interface to Linux seccomp filter",
    url: "https://github.com/seccomp/libseccomp/releases/download",
    version: "2.5.3",
    sha256sum: "59065c8733364725e9721ba48c3a99bbc52af921daf48df4b1e012fbc7b10a76",
};
```

and `include!("versions.rs")` it.

Reply:
I think the versions.rs is a great solution - I'll implement that as well. I might not be able to push all changes until Tuesday or later this week because I have some job interviews this week (if you're hiring let me know!).